Vdb 1318 CI/CD protection from out of order migrations

paytonrules opened the pull request:
This validates that every new migration comes after those in staging, so a staging deployment shouldn't fail. It also validates that the migrations are of a valid format.

paytonrules committed:
Fix typo

paytonrules committed:
Move flags specific to backfillUrns

These flags don't make sense in the checkMigrations command I'm
currently writing.

We may wish to, in the future, but anything that requires all these
params in one sub-package to specify the params once.

paytonrules committed:
Display error when new migraitons are out of order

Preliminary. This requires more testing, command line params and
integration with CI.

paytonrules committed:
Test filtering a list of files

To make it possible to test that little filter I had to add an
interface (fine) but a bunch of code to copy slice of objects into a
slice of interfaces (less fine). To make the GitHubfile structure match
the NamedFile interface I had to make its Name field lowercase (private). Also
fine, but then that meant it didn't marshal from JSON, so I had to add
an internal struct with exported fields. Whew.

paytonrules committed:
Extract NewMigrations func

And test it.

paytonrules committed:
Replace untested code with CheckMigrations function

Testing revealed I was mutating in place, which I did not want to do, so hooray!

paytonrules committed:
Check that migrations also start with timestamps

paytonrules committed:
Make the branch and repo configurable

You still have to point at Github though

paytonrules committed:
Log the URL you're retrieving migrations from

And the new migrations

paytonrules committed:
Move everything that's only used by backfillUrns

Anything that isn't shared is moved to backfill.go. This is kind of
misleading since all the code is in the same package anyway, but as part
of this I did need to move the PreRunPersistent in checkMigrationsCmd to
be PreRun so that both the logging and initGithubParams functions are called.

paytonrules committed:
Automate checking migrations via Travis

Add validatemigrationorder to Makefile. The name checkmigrations has other,
similar names in the Makefile related to generating migrations, so don't
use that.

Add an executable target to the Makefile, and depend on that for
validatemigrationorder. This is more make appropriate.

Make the build task always rebuild the executable, to match existing behavior.

paytonrules committed:
Return exit code 1 when check fails

paytonrules committed:
Stop build on failure

rmulhol commented on Makefile:
```diff
@@ -71,7 +71,17 @@ integrationtest: | $(GINKGO) $(LINT)
+# Build is really "clean/rebuild"
+.PHONY: build
 build:
+	rm vdb-mcd-transformers
```
Is this gonna run into trouble if you haven't run it before and the `vdb-mcd-transformers` binary doesn't exist?

paytonrules replied:
Yes! Needs a dash in front, nice catch.

rmulhol commented on cmd/backfillUrns.go:
```diff
@@ -34,8 +42,9 @@ var backfillUrnsCmd = &cobra.Command{
 	Short: "Backfill diffs for urns, looking up diffs based on associated events",
 	Long: `Fetch diffs when events indicate the state of an Urn changed at a given block.
-Optionally pass a starting block number to backfill since a given block.
+Optionally pass a staomm
```
?

paytonrules replied:
🦋 ⌨️

paytonrules replied:
Fixed - not sure how that go thtere.

rmulhol commented on cmd/checkMigrations.go:
```diff
@@ -0,0 +1,230 @@
+	newMigrations := NewMigrations(localMigrations, stagingMigrations)
+	logrus.Println("New Migrations are", newMigrations)
+
+	return CheckNewMigrations(stagingMigrations, newMigrations)
```
Would it potentially make sense to extract a lot of this code out of the `cmd` directory and test it there?

paytonrules replied:
Maybe - I debated it but decided against it because I can't imagine it being used elsewhere.

rmulhol commented on cmd/checkMigrations_test.go:
```diff
@@ -0,0 +1,130 @@
+			err := cmd.CheckNewMigrations(originalMigrations, newMigrations)
+
+			Expect(err).To(HaveOccurred())
+			Expect(newMigrations).To(Equal([]string{"20200429072515_newest.sql", "20200429072512_newest.sql"}))
```
What's the purpose of this expectation?

paytonrules replied:
To make sure newMigrations aren't changed in the function.

rmulhol commented on cmd/checkMigrations_test.go:
```diff
@@ -0,0 +1,130 @@
+
+			err := cmd.CheckNewMigrations(originalMigrations, newMigrations)
+
+			Expect(err).To(HaveOccurred())
```
nbd but could potentially be cool to assert `MatchError` if we want to be getting specific errors depending on the failure

paytonrules replied:
Can you match against a regex with MatchError? I don't like to match an entire string, since the tests get flaky, but I wouldn't object to matching the 'important' part.

elizabethengelman replied:
I've used [`MatchRegexp`](https://onsi.github.io/gomega/#matchregexpregexp-string-args-interface) before to match the important part of the error message, but I think that's matching the error string as opposed to the error itself. So, you'd need to do `Expect(err.Error()).To(MatchRegexp("important part"))` or something like that.

paytonrules replied:
That worked - perfect, thanks!

rmulhol commented on cmd/checkMigrations_test.go:
```diff
@@ -0,0 +1,130 @@
+			originalMigrations := []string{"20200429072513_last.sql"}
+			newMigrations := []string{"20200429072517_newest.sql", "2020042907251z_newest.sql"}
+
+			err := cmd.CheckNewMigrations(originalMigrations, newMigrations)
```
probably not worth changing this up for now but I'm wondering if this will need to be changed if/when we use `goose fix` to rename migrations

paytonrules replied:
It shouldn't - because that will number everything and should still be fine. 

`goose fix` could be problematic BTW - because it doesn't change the database for what migrations were already run. I learned that while starting this story.

rmulhol commented on cmd/checkMigrations.go:
```diff
@@ -0,0 +1,230 @@
+}
+
+func checkMigrations() error {
+	stagingMigrations, err := getGithubFileNames()
```
small nit but maybe this would be better named `remoteMigrations`, if we can configure the target branch to something other than staging

paytonrules replied:
👍

rmulhol commented on cmd/checkMigrations.go:
```diff
@@ -0,0 +1,230 @@
+	}
+
+	newMigrations := NewMigrations(localMigrations, stagingMigrations)
+	logrus.Println("New Migrations are", newMigrations)
```
does this look alright when it prints? Would have expected to see a `: ` after "are"

paytonrules replied:
Yup - println is smart:

```
{
  "level": "info",
  "msg": "New Migrations are [20200500000000_test.sql]",
  "time": "2020-06-10T09:14:00-05:00"
}
```

rmulhol commented on cmd/checkMigrations.go:
```diff
@@ -0,0 +1,230 @@
+		return err
+	}
+
+	localMigrations, err := getLocalMigrations()
```
nbd but I'm a fan of using different names for different errors since I think you could fail to check this error and the compiler wouldn't care - but that's not true if the error var has a different name than something that was assigned and checked previously

paytonrules replied:
👍

rmulhol commented on cmd/checkMigrations.go:
```diff
@@ -0,0 +1,230 @@
+	resp, err := http.Get(url)
+
+	if err != nil {
+		return []string{}, err
```
could be cool to use a wrapped error here to add context, e.g.
```
return []string{}, fmt.Errorf("error getting remote migrations: %w", err)
```

paytonrules replied:
👍

rmulhol reviewed (approved):
No blockers 👍

elizabethengelman reviewed (approved):
Looking good! 👍

paytonrules committed:
Add a dash to ignore errors on `rm` command

paytonrules committed:
Cleanup error messages

Use wrapped errors for clarification, and use differently errors instead
of reusing err.

paytonrules committed:
Update tests to handle exceptions
